### ISSUE TITLE
`ignore-patterns` incorrectly ignores some files

### ISSUE THREAD

**Issue by bersbersbers** [pylint-dev/pylint#9273]
### Bug description

```python
pylint incorrectly excludes files.


$ tree -a
.
├── .hidden
│   └── bug.py
└── bug.py
```
```


### Configuration

_No response_

### Command used

```shell
# working: ignores .hidden, but not bug.py
pylint --recursive=y --ignore-patterns="^\..+" . 

# bug: ignores everything!
pylint --recursive=y --ignore-patterns="^\." .
```

### Pylint output

```python
pylint --recursive=y --ignore-patterns="^\..+" . 
************* Module bug
bug.py:1:0: C0114: Missing module docstring (missing-module-docstring)
bug.py:1:5: E0602: Undefined variable 'f' (undefined-variable)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)

$ pylint --recursive=y --ignore-patterns="^\." .
$
```

### Expected behavior

Do not ignore `bug.py`

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.12.0 (main, Oct 25 2023, 10:28:41) [GCC 13.2.0]
```

### OS / Environment

Windows and Linux

### Additional dependencies

```python
None
```

**Comment by bersbersbers** [comment]
Same behavior as `pylint --recursive=y --ignore-patterns="^\..*" .`. Maybe the same as https://github.com/PyCQA/pycodestyle/issues/1217?
